persistence/src/test/java/com/epam/rft/atsy/persistence/repositories/PasswordHistoryRepositoryIT.java

```diff
@@ -0,0 +1,139 @@
+    private UserRepository userRepository;
+
+    @Test
+    public void findByUserEntityShouldNotFindPasswordHistoryForUserWithoutPasswordHistory() {
+        //Given
+        UserEntity user = this.userRepository.findByUserName(TEST_USERNAME_WITHOUT_PASSWORD_HISTORY);
+
+        //When
+        List<PasswordHistoryEntity> result = this.repository.findByUserEntity(user);
+
+        //Then
+        assertThat(result, notNullValue());
+        assertThat(result, empty());
+    }
+
+    @Test
+    public void findByUserEntityShouldFindSinglePasswordHistoryForUserWithSinglePasswordHistory() {
+        //Given
+        UserEntity user = this.userRepository.findByUserName(TEST_USERNAME_WITH_SINGLE_PASSWORD_HISTORY);
+        Date nearNow = currentDateMinus(5);
+
+        //When
+        List<PasswordHistoryEntity> result = this.repository.findByUserEntity(user);
+
+        //Then
+        assertThat(result, notNullValue());
+        assertThat(result.size(), is(1));
+        assertPasswordHistoryEntity(result.get(0), user, user.getUserPassword(), nearNow);
+    }
+
+    @Test
+    public void findByUserEntityShouldFindThreePasswordHistoryForUserWithThreePasswordHistory() {
+        //Given
+        UserEntity user = this.userRepository.findByUserName(TEST_USERNAME_WITH_THREE_PASSWORD_HISTORY);
+        Date nearNow = currentDateMinus(5);
+
+        //When
+        List<PasswordHistoryEntity> result = this.repository.findByUserEntity(user);
+
+        //Then
+        assertThat(result, notNullValue());
+        assertThat(result.size(), is(3));
+
+        assertPasswordHistoryEntity(result.get(0),
+                user,
+                user.getUserPassword(),
+                nearNow);
+
+        assertPasswordHistoryEntity(result.get(1),
+                user,
+                TEST_PASSWORD_1_FOR_USER_WITH_THREE_PASSWORD_HISTORY,
+                nearNow);
+
+        assertPasswordHistoryEntity(result.get(2),
+                user,
+                TEST_PASSWORD_2_FOR_USER_WITH_THREE_PASSWORD_HISTORY,
+                nearNow);
+
+    }
+
+    @Test
+    public void findOldestPasswordShouldNotFindPasswordHistoryForUserWithoutPasswordHistory() {
+        //When
+        PasswordHistoryEntity pwHistory = this.repository.findOldestPassword(TEST_USER_ID_WITHOUT_PASSWORD_HISTORY);
+
+        //Then
+        assertThat(pwHistory, nullValue());
+    }
+
+    @Test
+    public void findOldestPasswordShouldFindOldestPasswordHistoryForUserWithThreePasswordHistory() {
+        //Given
+        UserEntity user = this.userRepository.findByUserName(TEST_USERNAME_WITH_THREE_PASSWORD_HISTORY);
+        Date nearNow = currentDateMinus(5);
+
+        //When
+        PasswordHistoryEntity pwHistory = this.repository.findOldestPassword(TEST_USER_ID_WITH_THREE_PASSWORD_HISTORY);
+
+        //Then
+        assertThat(pwHistory, notNullValue());
+        assertPasswordHistoryEntity(pwHistory, user, ACTUAL_PASSWORD_FOR_USER_WITH_THREE_PASSWORD_HISTORY, nearNow);
```

Comment:
The test is failing on my machine with the following message:
> Expected: is "$2a$04$QSJkUouEDGfbWAtuxnlf/.Ajen6lviIhrNwKFPPZ.juRn6nLgvBi6"
>   but: was "password_history2"

The test data contains 3 password history entries in the database with change_date initialized with NOW() function. Most probably the date values are the same for all entries, so the query can return any of the 3 rows.

I suggest to prepare the test data to be specific which row is the oldest. You can use any MySQL date functions, eg: ADDDATE()